### PR TITLE
CLN: MultiIndex.get_value is a hive of scum and villainy

### PR DESCRIPTION
**Pull request by jbrockmendel**
This will in turn allow for simplification for other code that has built up kludges around MultiIndex.get_value's behavior

### CODE REVIEW

File: pandas/tests/resample/test_resample_api.py [hunk 1]
```diff
@@ -405,7 +405,7 @@ def test_agg_misc():
     ]
 
     # passed lambda
-    for t in cases:
+    for i, t in enumerate(cases):
```
Comment (MarcoGorelli) [inline pandas/tests/resample/test_resample_api.py]: Why? Where is this `i` used?

Comment (jbrockmendel) [reply pandas/tests/resample/test_resample_api.py]: this was in debugging, thought i had reverted it, thanks for pointing it out

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -2321,65 +2320,33 @@ def reindex(self, target, method=None, level=None, limit=None, tolerance=None):
...
         # Label-based
-        s = com.values_from_object(series)
-        k = com.values_from_object(key)
+        if not is_hashable(key) or is_iterator(key):
+            # We allow tuples if they are hashable, whereas other Index
+            #  subclasses require scalar.
+            # We have to explicitly exclude generators, as these are hashable.
+            raise InvalidIndexError(key)
 
         def _try_mi(k):
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: should think about making try_mi non-nested (just for cleanliness), give it a better name and type in the future.

**Review by jreback — APPROVED** [review]

File: pandas/core/indexes/multi.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/multi.py]: yah there's plenty more to do here, but this is a big step

**Pull request merged**